lib/recursive-methods.rb

```diff
@@ -1,49 +1,140 @@
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    # Input validation
+    raise ArgumentError if n < 0 
+    # Base case
+    return 1 if n == 0 
+    # Recursive case
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    # Input validation
+    return nil if s == nil
+    return '' if s.length == 0
+    # Base case
+    return s if s.length == 1
+    # Recursive case
+    return reverse(s[1..s.length-1]) + s[0]
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+    # Input validation
+    return nil if s == nil
+    return '' if s.length == 0    
+    i = 0
+    j = s.length-1
+    reverse_helper(s, i, j)
+    return s 
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def reverse_helper(s, i, j)
+    if i > j
+        return
+    end
+    s[i], s[j] = s[j], s[i]
+    reverse_helper(s, i+1, j-1)
+end
+
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    # base case
+    return 0 if n == 0
+    # recursive case
+    return 2 + bunny(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+    # input validation
+    return true if s.length == 0
+    # base case
+    return false if s.length == 1
+    if s[0] != '(' || s[-1] != ')'
+        return false
+    end
+    # recursive case
+    return nested(s[1..-2])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+    # input validation
+    return false if array.nil?
+    # base case
+    i = 0
+    return search_helper(array, i, value)
+end
+
+def search_helper(array, i, value)
+    # base case
+    if i == array.length
+        return false
+    end
+    if array[i] == value
+        return true
+    end
+    # recursive case
+    search_helper(array, i+1, value)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def is_palindrome(s)
-    raise NotImplementedError, "Method not implemented"
+    # input validation
+    return true if s.length == 0
+    i = 0
+    j = s.length-1
+    return palindrome_helper(i, j, s)
+end
+
+def palindrome_helper(i, j, s)
+    # base case
+    return true if j == i
+    return false if s[j] != s[i]
+    # recursive case
+    palindrome_helper(i+1, j-1, s)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(lg n)
+# Space complexity: O(lg n)
 def digit_match(n, m)
```

Comment:
I would say Log 10 n, instead of using the natural log, but you're pretty much right.  👍